File: Leetcode/Graphs/1162. As Far from Land as Possible.cpp
class Solution {
private:
    bool isValidPosition(int x, int y, int X, int Y) {
        return x >= 0 && y >= 0 && x < X && y < Y;
    }
public:
    int maxDistance(vector<vector<int>>& grid) {
        int N = grid.size(), M = grid[0].size(), maxDist = 0;
        vector <int> dS = {-1, 1, 0, 0};
        vector <vector <bool>> Visited(N, vector <bool> (M, false));
        queue <pair <int, int>> Queue;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (grid[i][j] == 1) {
                    Queue.push({i, j});
                    Visited[i][j] = true;
                    grid[i][j] = 0;
                } else grid[i][j] = 201;
            }
        }
        while (!Queue.empty()) {
            pair <int, int> Z = Queue.front();
            Queue.pop();
            for (int i = 0; i < 4; i++) {
                if (isValidPosition(Z.first + dS[i], Z.second + dS[3 - i], N, M)) {
                    if (!Visited[Z.first + dS[i]][Z.second + dS[3 - i]]) {
                        Visited[Z.first + dS[i]][Z.second + dS[3 - i]] = true;
                        Queue.push({Z.first + dS[i], Z.second + dS[3 - i]});
                        grid[Z.first + dS[i]][Z.second + dS[3 - i]] = min(grid[Z.first + dS[i]][Z.second + dS[3 - i]], 1 + grid[Z.first][Z.second]);
                    }
                }
            }
            maxDist = max(maxDist, grid[Z.first][Z.second]);
        }
        return maxDist == 0 ? -1 : maxDist;
    }
};

File: Leetcode/Graphs/1162. As Far from Land as Possible.py
import queue

class Solution:
    def maxDistance(self, grid: List[List[int]]) -> int:
        
        def isValidPos(x, y, X, Y):
            return x >= 0 and y >= 0 and x < X and y < Y
        
        N, M, Queue, maxDist, ds = len(grid), len(grid[0]), queue.Queue(), 0, [-1, 1, 0, 0]
        Visited = [[False for x in range(M)] for x in range(N)]
        
        for i in range(N):
            for j in range(M):
                if grid[i][j] == 1:
                    Queue.put((i, j))
                    grid[i][j] = 0
                    Visited[i][j] = True
                else:
                    grid[i][j] = 201
        
        while Queue.qsize() > 0:
            (x, y) = Queue.get()
            for k in range(4):
                if isValidPos(x + ds[k], y + ds[3 - k], N, M) and not Visited[x + ds[k]][y + ds[3 - k]]:
                    grid[x + ds[k]][y + ds[3 - k]] = min(grid[x + ds[k]][y + ds[3 - k]], 1 +  grid[x][y])
                    Queue.put((x + ds[k], y + ds[3 - k]))
                    Visited[x + ds[k]][y + ds[3 - k]] = True
            maxDist = max(maxDist, grid[x][y])
        
        return -1 if maxDist == 0 else maxDist